algorithms/maths/diffie_hellman_key_exchange.py

```diff
@@ -0,0 +1,183 @@
+
+"""
+For positive integer n and given integer a that satisfies gcd(a, n) = 1,
+a is the primitive root of n, if a's order k for n satisfies k = ϕ(n).
+Primitive roots of certain number may or may not be exist.
+If so, return empty list.
+"""
+
+def find_primitive_root(n):
+    if (n == 1):
+        return [0]
+        """ Exception Handeling :
+        0 is the only primitive root of 1 """
+    else:
+        phi = euler_totient(n)
+        p_root_list = []
+        """ It will return every primitive roots of n. """
+        for i in range (1, n):
+            if (math.gcd(i, n) != 1):
+                continue
+                """ To have order, a and n must be
+                relative prime with each other. """
+            else:
+                order = find_order(i, n)
+                if (order == phi):
+                    p_root_list.append(i)
+                else:
+                    continue
+        return p_root_list
+
+
+"""
+Diffie-Hellman key exchange is the method that enables
+two entities (in here, Alice and Bob), not knowing each other,
+to share common secret key through not-encrypted communication network.
+This method use the property of one-way function (discrete logarithm)
+For example, given a, b and n, it is easy to calculate x
+that satisfies (a^b) ≡ x (mod n).
+However, it is very hard to calculate x that satisfies (a^x) ≡ b (mod n).
+For using this method, large prime number p and its primitive root a must be given.
+"""
+
+def alice_private_key(p):
+    """Alice determine her private key
+    in the range of 1 ~ p-1.
+    This must be kept in secret"""
+    return randint(1, p-1)
+
+
+def alice_public_key(a_pr_k, a, p):
+    """Alice calculate her public key
+    with her private key.
+    This is open to public"""
+    return pow(a, a_pr_k) % p
+
+
+def bob_private_key(p):
+    """Bob determine his private key
+    in the range of 1 ~ p-1.
+    This must be kept in secret"""
+    return randint(1, p-1)
+
+
+def bob_public_key(b_pr_k, a, p):
+    """Bob calculate his public key
+    with his private key.
+    This is open to public"""
+    return pow(a, b_pr_k) % p
+    
+
+def alice_shared_key(b_pu_k, a_pr_k, p):
+    """ Alice calculate secret key shared with Bob,
+    with her private key and Bob's public key.
+    This must be kept in secret"""
+    return pow(b_pu_k, a_pr_k) % p
+
+
+def bob_shared_key(a_pu_k, b_pr_k, p):
+    """ Bob calculate secret key shared with Alice,
+    with his private key and Alice's public key.
+    This must be kept in secret"""
+    return pow(a_pu_k, b_pr_k) % p
+
+
+def diffie_hellman_key_exchange(a, p):
+    if (prime_check(p) == False):
+        print("%d is not a prime number" % p)
+        return -1
+        """p must be large prime number"""
+    else:
+        try:
+            p_root_list = find_primitive_root(p)
+            p_root_list.index(a)
+        except ValueError as e:
+            print("%d is not a primitive root of %d" % (a, p))
+            return -1
```

Comment:
please raise error in this case so we can only return bool type.
If a function returns two types (-1, bool) the caller needs to branch out by the return type.

Reply:
[Edit] For primitive root case, code now return False, for consistency